### ISSUE TITLE
Question regarding t_max and d estimation

### ISSUE THREAD

**Issue by DanPli** [konstmish/prodigy#6]
Hello :)

First, thank you both for your work, absolutely loving d-adaptation and now prodigy.

My question: 
(Training a Stable Diffusion LoRa.) 
Using CosineAnnealingLR and setting a t_max that is not t_max=n_epochs, will this throw off Prodigy's d estimation completely ? 

Generally wondering if scaling the restarts using t_max is a bad idea. I have a training setup coming up that I can't avoid having many steps and fewer epochs, and I was wondering if I could scale the restarts by setting t_max in that case to half or even n_epochs/4 to get more restarts for the entire training without a detrimental effect on prodigy in doing so. 

I already used an approach of t_max = 4 * n_epochs to have more iterations between restarts on another training with very few steps/many epochs, and based on those results I am in doubt if this approach is futile and if I am better off using cosine with restarts instead. Or if I should purely tune d_coef and leave the scheduler t_max at n_epochs.

If you could give me a hint on this topic I would really appreciate it.

Thank you :)

**Comment by konstmish** [comment]
Hi @DanPli,

Thanks for the interest in our work! We'd be happy if the method is helpful to you, so let us know if there are any other things you need help with.

To use cosine annealing with restarts, I think it's best to pass the argument `safeguard_warmup=True` to the method, which makes sure that for small stepsizes at the end of a cosine cycle, Prodigy doesn't try to increase the stepsize to compensate for a small stepsize. We used `safeguard_warmup=True` for warm-up epochs (without it, the Prodigy was estimating a stepsize that was too large), but it might actually be useful for restarts as well. I don't think `d_coef` needs to be changed for this, but I have low confidence since we haven't tested restarts ourselves.
